do not emit None mid-stream during parse

panicbit opened the pull request:
<!--
if this PR closes one or more issues, you can automatically link the PR with
them by using one of the [*linking keywords*](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword), e.g.
- this PR should close #issue 

you can also mention related issues, PRs or discussions!
-->

# Description
<!--
Thank you for improving Nushell. Please, check our [contributing guide](../CONTRIBUTING.md) and talk to the core team before making major changes.

Description of your pull request goes here. **Provide examples and/or screenshots** if your changes affect the user experience.
-->
Currently `parse` acts like a `.filter` over an iterator, except that it emits `None` for elements that can't be parsed. This causes consumers of the adapted iterator to stop iterating too early. The correct behaviour is to keep pulling the inner iterator until either the end of it is reached or an element can be parsed.

- this PR should close #9906 

# User-Facing Changes
<!-- List of all changes that impact the user experience here. This helps us keep track of breaking changes. -->
List streams won't be truncated anymore after the first parse failure.

# Tests + Formatting
<!--
Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- [x] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [x] `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect -A clippy::result_large_err` to check that you're using the standard code style
- [-] `cargo test --workspace` to check that all tests pass

- `cargo run -- -c "use std testing; testing run-tests --path crates/nu-std"` to run the tests for the standard library

> **Note**
> from `nushell` you can also use the `toolkit` as follows
> ```bash
> use toolkit.nu  # or use an `env_change` hook to activate it automatically
> toolkit check pr
> ```
-->
- [x] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [x] `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect -A clippy::result_large_err` to check that you're using the standard code style
- [x] `cargo test --workspace` to check that all tests pass
  - 11 tests fail, but the same 11 tests fail on main as well
 - [x] `cargo run -- -c "use std testing; testing run-tests --path crates/nu-std"` to run the tests for the standard library

# After Submitting
<!-- If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date. -->

fdncred commented:
nice work @panicbit! I'll have to test this out tomorrow.

amtoine reviewed (changes requested):
it looks like it's solving some issues with `parse`, nice job @panicbit :pray: 

i see
```nushell
let matches = [a b c] | str join "\n" | lines | parse -r '([ac])'; print $matches
```
gives a wrong
```nushell
╭───┬──────────╮
│ # │ capture0 │
├───┼──────────┤
│ 0 │ a        │
╰───┴──────────╯
```
on latest `main` but looks fixed on the tip of this branch :partying_face: 
```nushell
╭───┬──────────╮
│ # │ capture0 │
├───┼──────────┤
│ 0 │ a        │
│ 1 │ c        │
╰───┴──────────╯
```

could you add a test for this to make sure it does not happen again? :yum:

amtoine commented:
i also have this more complex command with a bunch of lines, regex-like characters, the whole set of complicated stuff you know :smirk:
```nushell
tmux show -g | lines | parse "status-format[1] {format}"
```

on latest `main`, it's not able to parse it and returns an empty list...
this is the same issue i think!


on the tip of this branch, it gives the expected result :partying_face:
```nushell
╭───┬─────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ # │                                             format                                                      │
├───┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ 0 │ "#[align=centre]#{P:#{?pane_active,#[reverse],}#{pane_index}[#{pane_width}x#{pane_height}]#[default] }" │
╰───┴─────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

amtoine commented:
and wow, i can not reproduce #8352 either on the tip of this branch, amazing :star_struck:

panicbit commented:
I have added a test, however, I noticed odd behaviour with the tests that I do not quite understand.
`cargo test --workspace` does not run the test I added, only `cargo test --workspace parse_does_not_truncate_list_streams` does.

fdncred commented on crates/nu-command/tests/commands/parse.rs:
```diff
@@ -197,4 +197,21 @@ mod regex {
+                r#"
+                    [a b c]
+                    | each {|x| $x}
+                    | parse --regex "[ac]"
```
Shouldn't this be `| parse --regex "([ac])"`? I think it works either way but `[ac]` produces no usable output since there isn't a capture group.

panicbit replied:
Yeah, it doesn't really matter for the test, but I'll add the parens since it makes the test slighty more understandable when looking at the output of that parse manually.

fdncred commented:
This looks to me like it works as expected. Nice work!

amtoine commented on crates/nu-command/tests/commands/parse.rs:
```diff
@@ -197,4 +197,21 @@ mod regex {
+
+            assert_eq!(actual.out, "2");
+        })
+    }
```
maybe it would need to be formatted with `cargo fmt --all` but i suggest simplifying the test in the following way :yum: 
```suggestion
    #[test]
    fn parse_does_not_truncate_list_streams() {
        let actual = nu!(pipeline(
            r#"
                [a b c]
                | each {|x| $x}
                | parse --regex "[ac]"
                | length
            "#
        ));

        assert_eq!(actual.out, "2");
    }
```

amtoine replied:
the other tests in that file might be incorrect as well...
this is part of https://github.com/nushell/nushell/issues/8670 :thinking:

panicbit replied:
Good catch!

amtoine replied:
thanks for the change :pray:

fdncred commented:
Thanks!

amtoine commented:
thanks so much @panicbit for fixing this :pray: 
`parse` feels usable now, loving this :blush:
